presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryMetadata.java

```diff
@@ -0,0 +1,217 @@
+    {
+        MemoryTableHandle memoryTableHandle = checkType(tableHandle, MemoryTableHandle.class, "tableHandle");
+        return memoryTableHandle.getColumnHandles().stream()
+                .collect(toMap(MemoryColumnHandle::getName, column -> column));
+    }
+
+    @Override
+    public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle)
+    {
+        MemoryColumnHandle memoryColumnHandle = checkType(columnHandle, MemoryColumnHandle.class, "columnHandle");
+        return memoryColumnHandle.toColumnMetadata();
+    }
+
+    @Override
+    public Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSession session, SchemaTablePrefix prefix)
+    {
+        return tables.values().stream()
+                .filter(table -> prefix.matches(table.toSchemaTableName()))
+                .collect(toMap(MemoryTableHandle::toSchemaTableName, handle -> handle.toTableMetadata().getColumns()));
+    }
+
+    @Override
+    public void dropTable(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        MemoryTableHandle handle = checkType(tableHandle, MemoryTableHandle.class, "tableHandle");
+        Long tableId = tableIds.remove(handle.getTableName());
+        if (tableId != null) {
+            tables.remove(tableId);
+        }
+    }
+
+    @Override
+    public void renameTable(ConnectorSession session, ConnectorTableHandle tableHandle, SchemaTableName newTableName)
+    {
+        MemoryTableHandle oldTableHandle = checkType(tableHandle, MemoryTableHandle.class, "tableHandle");
+        MemoryTableHandle newTableHandle = new MemoryTableHandle(
+                oldTableHandle.getConnectorId(),
+                oldTableHandle.getSchemaName(),
+                newTableName.getTableName(),
+                oldTableHandle.getTableId(),
+                oldTableHandle.getColumnHandles(),
+                oldTableHandle.getActiveTableIds()
+        );
+        tableIds.remove(oldTableHandle.getTableName());
+        tableIds.put(newTableName.getTableName(), oldTableHandle.getTableId());
+        tables.remove(oldTableHandle.getTableId());
+        tables.put(oldTableHandle.getTableId(), newTableHandle);
+    }
+
+    @Override
+    public void createTable(ConnectorSession session, ConnectorTableMetadata tableMetadata)
+    {
+        ConnectorOutputTableHandle outputTableHandle = beginCreateTable(session, tableMetadata, Optional.empty());
+        finishCreateTable(session, outputTableHandle, ImmutableList.of());
+    }
+
+    @Override
+    public ConnectorOutputTableHandle beginCreateTable(ConnectorSession session, ConnectorTableMetadata tableMetadata, Optional<ConnectorNewTableLayout> layout)
+    {
+        long nextId = nextTableId.getAndIncrement();
+        Set<Long> activeTableIds = new HashSet(tableIds.values());
+        activeTableIds.add(nextId);
```

Comment:
indeed.

However there is one bug with transactions. Scenario:

1. beginCreateTable() for table 100
2. beginCreateTable() for table 101
3. inserting some pages to 100 and 101
4. finishCreateTable() for table 101
5. beginCreateTable() for table 102

now any write to table 102 will remove any pages of table 100, because table 100 has not been added to `activeTableIds` yet. I will address this issue later this week or next week.